File: web/components/push.go
package components

import (
	"context"
	"errors"

	"github.com/maxence-charriere/go-app/v10/pkg/app"
)

// ErrPushUnsupported is returned by a PushSubscriber when the browser exposes no
// usable Push API (so the toggle renders an unsupported state instead of an
// error).
var ErrPushUnsupported = errors.New("push not supported")

// ErrPushDenied is returned when the user denied the browser notification
// permission prompt, so the toggle can explain rather than retry blindly.
var ErrPushDenied = errors.New("push permission denied")

// PushSubscriber abstracts the browser Push API behind an interface so the
// subscription flow is testable without a real browser (TESTING.md: mock the
// browser Notification/Push APIs behind an abstraction). The production
// implementation drives navigator.serviceWorker / PushManager via the go-app JS
// bridge; tests inject a mock that returns canned subscriptions.
//
// No business logic lives here: it only translates between the browser Push API
// and the PushSubscription shape Rails stores.
type PushSubscriber interface {
	// Supported reports whether the Push API is usable in this environment.
	Supported() bool

	// Subscribe requests notification permission (if needed) and subscribes to
	// push with the given application server (public VAPID) key, returning the
	// resulting browser subscription. It returns ErrPushDenied when the user
	// declines permission.
	Subscribe(ctx context.Context, vapidPublicKey string) (PushSubscription, error)

	// CurrentEndpoint returns the endpoint of the active browser subscription,
	// or "" when there is none.
	CurrentEndpoint(ctx context.Context) (string, error)

	// Unsubscribe cancels the active browser push subscription.
	Unsubscribe(ctx context.Context) error
}

// pushUIState is the rendered state of the push toggle.
type pushUIState int

const (
	pushUnknown     pushUIState = iota // initial / still determining
	pushUnsupported                    // browser cannot do push
	pushOff                            // supported, not subscribed
	pushOn                             // subscribed
	pushBusy                           // a subscribe/unsubscribe is in flight
	pushFailed                         // last action failed
)

// PushToggle is the push-subscription control. It reads the public VAPID key
// from Rails, subscribes via the browser Push API, and posts the resulting
// subscription to Rails (POST /api/push_subscription) for the daily digest. The
// owner can unsubscribe, which cancels the browser subscription and removes it
// from Rails (DELETE /api/push_subscription). The browser Push API is reached
// through the Pusher abstraction so the flow is testable without a browser.
type PushToggle struct {
	app.Compo

	// Client is the Rails client; tests inject a mock and it defaults to the
	// same-origin HTTP client on mount.
	Client RailsClient
	// Pusher is the browser Push API abstraction; tests inject a mock. When nil
	// it defaults to the live browser subscriber on mount.
	Pusher PushSubscriber

	state    pushUIState
	endpoint string
	errMsg   string
}

func (t *PushToggle) OnMount(ctx app.Context)     { t.start(ctx) }
func (t *PushToggle) OnPreRender(ctx app.Context) { t.start(ctx) }

func (t *PushToggle) start(ctx app.Context) {
	if t.Client == nil {
		t.Client = NewRailsClient()
	}
	if t.Pusher == nil {
		t.Pusher = NewBrowserPusher()
	}
	t.refresh(ctx)
}

// refresh determines the current subscription state from the browser.
func (t *PushToggle) refresh(ctx app.Context) {
	reqCtx := ctx.Context
	pusher := t.Pusher
	ctx.Async(func() {
		if !pusher.Supported() {
			ctx.Dispatch(func(ctx app.Context) {
				t.state = pushUnsupported
				ctx.Update()
			})
			return
		}
		endpoint, err := pusher.CurrentEndpoint(reqCtx)
		ctx.Dispatch(func(ctx app.Context) {
			if err != nil {
				t.state = pushFailed
				t.errMsg = pushErrorMessage(err)
				ctx.Update()
				return
			}
			t.endpoint = endpoint
			t.state = initialPushState(true, endpoint)
			ctx.Update()
		})
	})
}

// initialPushState maps supported/endpoint facts to the rendered toggle state.
// Pure so the gating is unit-testable without the engine.
func initialPushState(supported bool, endpoint string) pushUIState {
	if !supported {
		return pushUnsupported
	}
	if endpoint != "" {
		return pushOn
	}
	return pushOff
}

// enable is wired to the subscribe button's OnClick. It runs only on an
// explicit user action: fetch the VAPID key, subscribe via the browser, then
// store the subscription in Rails.
func (t *PushToggle) enable(ctx app.Context, _ app.Event) {
	if t.state == pushBusy {
		return
	}
	t.state = pushBusy
	t.errMsg = ""
	ctx.Update()
	reqCtx := ctx.Context
	ctx.Async(func() {
		sub, err := t.doSubscribe(reqCtx)
		ctx.Dispatch(func(ctx app.Context) {
			t.applySubscribe(sub, err)
			ctx.Update()
		})
	})
}

// doSubscribe performs the full subscribe path: read the VAPID key, subscribe in
// the browser, and persist to Rails. Returned synchronously so tests can drive
// the whole flow without the go-app engine.
func (t *PushToggle) doSubscribe(ctx context.Context) (PushSubscription, error) {
	key, err := t.Client.VAPIDPublicKey(ctx)
	if err != nil {
		return PushSubscription{}, err
	}
	if key == "" {
		return PushSubscription{}, ErrPushUnsupported
	}
	sub, err := t.Pusher.Subscribe(ctx, key)
	if err != nil {
		return PushSubscription{}, err
	}
	if err := t.Client.Subscribe(ctx, sub); err != nil {
		return PushSubscription{}, err
	}
	return sub, nil
}

// applySubscribe records the outcome of a subscribe attempt. Split from the
// engine plumbing so the state transitions are unit-testable.
func (t *PushToggle) applySubscribe(sub PushSubscription, err error) {
	if err != nil {
		if errors.Is(err, ErrPushUnsupported) {
			t.state = pushUnsupported
			return
		}
		t.state = pushFailed
		t.errMsg = pushErrorMessage(err)
		return
	}
	t.endpoint = sub.Endpoint
	t.state = pushOn
}

// disable is wired to the unsubscribe button's OnClick. It cancels the browser
// subscription and removes it from Rails.
func (t *PushToggle) disable(ctx app.Context, _ app.Event) {
	if t.state == pushBusy {
		return
	}
	t.state = pushBusy
	t.errMsg = ""
	ctx.Update()
	reqCtx := ctx.Context
	endpoint := t.endpoint
	ctx.Async(func() {
		err := t.doUnsubscribe(reqCtx, endpoint)
		ctx.Dispatch(func(ctx app.Context) {
			t.applyUnsubscribe(err)
			ctx.Update()
		})
	})
}

// doUnsubscribe cancels the browser subscription then removes it from Rails.
func (t *PushToggle) doUnsubscribe(ctx context.Context, endpoint string) error {
	if err := t.Pusher.Unsubscribe(ctx); err != nil {
		return err
	}
	return t.Client.Unsubscribe(ctx, endpoint)
}

func (t *PushToggle) applyUnsubscribe(err error) {
	if err != nil {
		t.state = pushFailed
		t.errMsg = pushErrorMessage(err)
		return
	}
	t.endpoint = ""
	t.state = pushOff
}

// pushErrorMessage maps a push flow error to a user-facing message.
func pushErrorMessage(err error) string {
	switch {
	case errors.Is(err, ErrPushDenied):
		return "Notifications were blocked. Enable them in your browser settings to get the daily digest."
	case IsUnauthorized(err):
		return "Your session expired. Please sign in again."
	default:
		return "Could not update push notifications. Please try again."
	}
}

func (t *PushToggle) Render() app.UI {
	return app.Div().Class("push-toggle").Body(
		app.H2().Text("Push notifications"),
		app.P().Class("push-toggle-note").
			Text("Get the daily job digest as a push notification on this device."),
		renderPushControl(t),
		app.If(t.state == pushFailed, func() app.UI {
			return app.P().Class("push-toggle-error").Text(t.errMsg)
		}),
	)
}

func renderPushControl(t *PushToggle) app.UI {
	switch t.state {
	case pushUnsupported:
		return app.P().Class("push-toggle-unsupported").
			Text("Push notifications aren't available in this browser. Install the app to your home screen on iOS 16.4+.")
	case pushOn:
		return app.Div().Class("push-toggle-control").Body(
			app.Span().Class("push-toggle-status push-toggle-status-on").Text("On"),
			app.Button().
				Class("push-toggle-disable").
				OnClick(t.disable).
				Text("Turn off notifications"),
		)
	case pushBusy:
		return app.Button().Class("push-toggle-busy").Disabled(true).Text("Working…")
	default: // pushOff, pushUnknown, pushFailed
		return app.Button().
			Class("push-toggle-enable").
			OnClick(t.enable).
			Text("Turn on notifications")
	}
}

File: client/src/lib/push.ts
/**
 * The browser Push API, ported from `web/components/push_browser.go` (see docs/GO_MIGRATION.md).
 *
 * The Go build reached the PushManager through go-app's JS bridge: every call went out as
 * `app.Window().Get(...)`, and because Go has no `await`, each returned promise was funnelled
 * back through a hand-written adapter that allocated two `app.FuncOf` callbacks, `Release`d them,
 * and blocked a goroutine on a channel. That adapter — and the `app.Value` marshalling around it,
 * which serialized the subscription with `JSON.stringify` only to `json.Unmarshal` it again — is
 * the bulk of the 157 lines this module replaces. In TypeScript the browser API is just the
 * browser API, so what is left is the part that was always real: the four operations the toggle
 * needs, and the two error kinds it renders differently.
 *
 * ## Nothing here subscribes on its own
 *
 * `subscribe` is only ever reached from an explicit click (`components/push-toggle.tsx`). That is
 * not merely a UI convention: requesting notification permission without a user gesture is how a
 * browser decides to block the origin permanently, and `AGENTS.md` requires the owner to opt in.
 * `currentEndpoint` is the only call made on mount, and it reads existing state without prompting.
 *
 * ## The seam
 *
 * `PushEnvironment` is the narrow slice of the browser this module touches — service-worker
 * readiness, the permission prompt, and a feature check. Injecting it is what lets the tests run
 * the real flow against a **mocked PushManager**: no notification is ever requested and no push is
 * ever sent. It is the same shape of seam as `OpenrouterClient`'s `http:` transport in Rails and
 * the `RailsClient` mock the Go screens used.
 *
 * Rails is not called from here at all. Reading the VAPID key and persisting the subscription are
 * the component's job, so this module stays a pure translation of the browser API into the
 * `PushSubscription` shape Rails stores.
 */
import { isUnauthorized } from "../api/errors";
import { PushSubscriptionSchema } from "../api/schemas";
import type { PushSubscription as PushSubscriptionPayload } from "../api/schemas";

/* -------------------------------------------------------------------------- */
/* Errors                                                                      */
/* -------------------------------------------------------------------------- */

/**
 * The browser exposes no usable Push API, so the toggle renders its unsupported guidance
 * instead of an error. Ported from `ErrPushUnsupported`.
 *
 * Also raised when Rails has no VAPID key configured: from the owner's side "this device cannot
 * receive push" is the same fact whether the browser or the server is missing the capability.
 */
export class PushUnsupportedError extends Error {
  override readonly name = "PushUnsupportedError";

  constructor(message = "push not supported") {
    super(message);
  }
}

/** The owner declined the notification permission prompt. Ported from `ErrPushDenied`. */
export class PushDeniedError extends Error {
  override readonly name = "PushDeniedError";

  constructor(message = "push permission denied") {
    super(message);
  }
}

/* -------------------------------------------------------------------------- */
/* The browser seam                                                            */
/* -------------------------------------------------------------------------- */

/** The subset of `PushSubscription` (the browser object) this module reads. */
export interface BrowserPushSubscription {
  readonly endpoint: string;
  toJSON(): { endpoint?: string | null | undefined; keys?: Record<string, string> | undefined };
  unsubscribe(): Promise<boolean>;
}

/** The subset of `PushManager` this module drives. */
export interface PushManagerLike {
  subscribe(options: {
    userVisibleOnly: boolean;
    applicationServerKey: string;
  }): Promise<BrowserPushSubscription>;
  getSubscription(): Promise<BrowserPushSubscription | null>;
}

/** The subset of `ServiceWorkerRegistration` this module reads. */
export interface PushRegistration {
  readonly pushManager: PushManagerLike;
}

/**
 * The browser capabilities the push flow needs. A real `ServiceWorkerRegistration` and
 * `PushManager` satisfy these structurally, so the production environment below is a thin
 * forwarder and tests supply a fake with no DOM at all.
 */
export interface PushEnvironment {
  /** Whether the Push API is usable here. False in jsdom, in SSR, and in unsupported browsers. */
  supported(): boolean;
  /** `navigator.serviceWorker.ready`. */
  ready(): Promise<PushRegistration>;
  /** `Notification.requestPermission()`. Shows the browser prompt — user gesture only. */
  requestPermission(): Promise<NotificationPermission>;
}

/** What the toggle calls. Ported from the Go `PushSubscriber` interface. */
export interface PushSubscriber {
  /** Whether the Push API is usable in this environment. */
  supported(): boolean;
  /**
   * Requests notification permission if needed and subscribes with the given application server
   * (public VAPID) key. Rejects with `PushDeniedError` when the owner declines and
   * `PushUnsupportedError` when the Push API is missing.
   */
  subscribe(vapidPublicKey: string): Promise<PushSubscriptionPayload>;
  /** The active browser subscription's endpoint, or `""` when there is none. */
  currentEndpoint(): Promise<string>;
  /** Cancels the active browser subscription, if there is one. */
  unsubscribe(): Promise<void>;
}

/**
 * The live browser environment.
 *
 * Every lookup is guarded because this runs in three places that are not browsers: jsdom under
 * Vitest, a prerender, and any browser without the Push API (Safari before iOS 16.4, and iOS at
 * all until the app is installed to the home screen). All three answer "unsupported", which is a
 * rendered state rather than an error — the same thing go-app's server-side JS shim produced.
 */
export function browserPushEnvironment(): PushEnvironment {
  return {
    supported: pushApiAvailable,
    ready: () => globalThis.navigator.serviceWorker.ready,
    requestPermission: () => globalThis.Notification.requestPermission(),
  };
}

function pushApiAvailable(): boolean {
  try {
    const nav: Navigator | undefined = globalThis.navigator;
    if (!nav?.serviceWorker) return false;
    return "PushManager" in globalThis && typeof globalThis.Notification !== "undefined";
  } catch {
    // Touching `navigator` can throw outside a document context; that is also "unsupported".
    return false;
  }
}

/**
 * The production `PushSubscriber`, over whatever environment it is handed.
 *
 * Ordering inside `subscribe` is deliberate and matches `push_browser.go`: permission first, then
 * service-worker readiness, then the subscription. Asking for permission before awaiting `ready`
 * keeps the prompt inside the click's user-gesture window, which some browsers require.
 */
export function createPushSubscriber(
  environment: PushEnvironment = browserPushEnvironment(),
): PushSubscriber {
  return {
    supported: () => environment.supported(),

    async subscribe(vapidPublicKey: string): Promise<PushSubscriptionPayload> {
      if (!environment.supported()) throw new PushUnsupportedError();

      const permission = await environment.requestPermission();
      if (permission !== "granted") throw new PushDeniedError();

      const registration = await environment.ready();
      // The key is passed as the base64url string Rails serves, exactly as the Go build did.
      // `PushSubscriptionOptionsInit.applicationServerKey` accepts a DOMString and decodes it,
      // so there is no byte-array conversion to get wrong.
      const subscription = await registration.pushManager.subscribe({
        userVisibleOnly: true,
        applicationServerKey: vapidPublicKey,
      });
      return readSubscription(subscription);
    },

    async currentEndpoint(): Promise<string> {
      if (!environment.supported()) return "";
      const registration = await environment.ready();
      const subscription = await registration.pushManager.getSubscription();
      // Read the endpoint off the object rather than through `readSubscription`: this runs on
      // every mount, and a subscription whose keys the browser will not re-expose must still
      // report "on" rather than flipping the toggle to a failure the owner cannot act on.
      return subscription?.endpoint ?? "";
    },

    async unsubscribe(): Promise<void> {
      if (!environment.supported()) return;
      const registration = await environment.ready();
      const subscription = await registration.pushManager.getSubscription();
      if (!subscription) return;
      await subscription.unsubscribe();
    },
  };
}

/**
 * Reads a browser subscription into the payload Rails stores.
 *
 * `toJSON()` is the source, as in `decodeSubscription`: it is the only way to get at the
 * `p256dh` and `auth` encryption keys, which are not properties of the subscription object.
 * Unlike Go's `json.Unmarshal`, a missing endpoint or missing keys is rejected rather than
 * zero-valued — an empty key posted to Rails produces a stored subscription that can never be
 * encrypted to, and the resulting silence looks exactly like a working subscription.
 */
export function readSubscription(subscription: BrowserPushSubscription): PushSubscriptionPayload {
  const json = subscription.toJSON();
  const parsed = PushSubscriptionSchema.safeParse({
    endpoint: json.endpoint ?? subscription.endpoint,
    keys: { p256dh: json.keys?.["p256dh"], auth: json.keys?.["auth"] },
  });
  if (!parsed.success) {
    throw new Error("browser push subscription is missing its endpoint or encryption keys");
  }
  return parsed.data;
}

/* -------------------------------------------------------------------------- */
/* Rendered state                                                              */
/* -------------------------------------------------------------------------- */

/**
 * The toggle's rendered state. Ported from `pushUIState`, with `denied` split out of the Go
 * build's single `pushFailed`: a blocked permission is not a transient failure — retrying cannot
 * fix it — so it is its own state, distinct from `unsupported`, `on`, and `off`. It renders the
 * same markup `pushFailed` did (the enable control plus `.push-toggle-error`), only with the
 * guidance message pinned, so no new stylesheet rule is involved.
 */
export type PushUiState = "unknown" | "unsupported" | "off" | "on" | "busy" | "denied" | "failed";

/** Whether a state shows the error paragraph. */
export function hasPushError(state: PushUiState): boolean {
  return state === "denied" || state === "failed";
}

/**
 * Maps supported/endpoint facts to the rendered state. Ported from `initialPushState`, and pure
 * for the same reason: it is the gate that decides whether the owner sees "on", and it should be
 * assertable without a browser.
 */
export function initialPushState(supported: boolean, endpoint: string): PushUiState {
  if (!supported) return "unsupported";
  return endpoint !== "" ? "on" : "off";
}

/** The state a failed subscribe/unsubscribe lands in. */
export function pushErrorState(error: unknown): PushUiState {
  if (error instanceof PushUnsupportedError) return "unsupported";
  if (error instanceof PushDeniedError) return "denied";
  return "failed";
}

/** `pushErrorMessage`, unchanged: the three owner-facing strings, byte for byte. */
export function pushErrorMessage(error: unknown): string {
  if (error instanceof PushDeniedError) {
    return "Notifications were blocked. Enable them in your browser settings to get the daily digest.";
  }
  if (isUnauthorized(error)) {
    return "Your session expired. Please sign in again.";
  }
  return "Could not update push notifications. Please try again.";
}
